transaction-view/src/resolved_transaction_view.rs

```diff
@@ -0,0 +1,217 @@
+    /// addresses, create a resolved transaction view.
+    pub fn new(
+        view: TransactionView<true, D>,
+        resolved_addresses: LoadedAddresses,
+        reserved_account_keys: &HashSet<Pubkey>, // why does this not use ahash at least?!
+    ) -> Self {
+        let writable_cache =
+            Self::cache_is_writable(&view, &resolved_addresses, reserved_account_keys);
+        Self {
+            view,
+            resolved_addresses,
+            writable_cache,
+        }
+    }
+
+    /// Helper function to check if an address is writable,
+    /// and cache the result.
+    /// This is done so we avoid recomputing the expensive checks each time we call
+    /// `is_writable` - since there is more to it than just checking index.
+    fn cache_is_writable(
+        view: &TransactionView<true, D>,
+        resolved_addresses: &LoadedAddresses,
+        reserved_account_keys: &HashSet<Pubkey>, // why does this not use ahash at least?!
+    ) -> Vec<bool> {
+        // Build account keys so that we can iterate over and check if
+        // an address is writable.
+        let account_keys = AccountKeys::new(view.static_account_keys(), Some(resolved_addresses));
+
+        let mut is_writable_cache = Vec::with_capacity(account_keys.len());
+        let num_static_accounts = usize::from(view.num_static_account_keys());
+        let num_writable_lookup_accounts = usize::from(view.total_writable_lookup_accounts());
+        let num_signed_accounts = usize::from(view.num_required_signatures());
+        let num_unsigned_accounts = num_static_accounts.wrapping_sub(num_signed_accounts);
+        let num_writable_unsigned_accounts =
+            num_unsigned_accounts.wrapping_sub(usize::from(view.num_readonly_unsigned_accounts()));
+        let num_writable_signed_accounts =
+            num_signed_accounts.wrapping_sub(usize::from(view.num_readonly_signed_accounts()));
+
+        let is_upgradable_loader_present = account_keys
+            .iter()
+            .any(|key| bpf_loader_upgradeable::ID == *key);
+
+        for (index, key) in account_keys.iter().enumerate() {
+            let is_requested_write = {
+                // If the account is a resolved address, check if it is writable.
+                if index >= num_static_accounts {
+                    let loaded_address_index = index.wrapping_sub(num_static_accounts);
+                    loaded_address_index < num_writable_lookup_accounts
+                } else if index >= num_signed_accounts {
+                    let unsigned_account_index = index.wrapping_sub(num_signed_accounts);
+                    unsigned_account_index < num_writable_unsigned_accounts
+                } else {
+                    index < num_writable_signed_accounts
+                }
+            };
+
+            // If the key is reserved it cannot be writable.
+            is_writable_cache[index] = is_requested_write && !reserved_account_keys.contains(key);
```

Comment:
sdk's version has `&& !message.demote_program_id(i)`, is this deprecated?

Reply:
There is a separate loop under this one which will handle that demotion.
But it's implemented in a much more efficient way.

The `sdk` version is `O(N*K)` where `N` is number of instructions and `K` is number of keys. This is because for each account key index, it loops over all instructions to see if there are any that use it as a program.

This version simply loops over the instructions, and then demotes indexes that are used as programs (if the upgradable loader is **not** present).
So it is only `O(N)`.